### DISCUSSION TITLE
C# classes in Skin editor's JSON file, Can this be vulnerable?

### DISCUSSION THREAD

**Discussion by SK-Fast** [ppy/osu#20837] [Q&A]
First of all. I'm not a C# programmer, I'm pretty much a Web developer with a bit of knowledge of C#. I'm just going through trial & error to see how osu!lazer reacts, Couldn't check the code and debug.

So today I found out that you can edit the Skin editor's JSON file `MainHUDComponents.json`, and you can change the "Type" field to any existing classes in `osu.Game.xml`, even the one that does not include in the skin editor's menu.

Here's example of what I'm talking about editing:
```json
  {
    "Type": "osu.Game.Screens.Menu.OsuLogo, osu.Game, Version=2022.1008.2.0, Culture=neutral, PublicKeyToken=null",
    "Position": {
      "x": 0,
      "y": 0
    },
    "Rotation": 0.0,
    "Scale": {
      "x": 0.5,
      "y": 0.5
    },
    "Anchor": 18,
    "Origin": 17,
    "UsesFixedAnchor": false,
    "Settings": {},
    "Children": []
  }
```
![the osu logo appears](https://user-images.githubusercontent.com/60780448/196885269-99bac03e-68bd-4009-b49d-2422f9a5116c.png)
_the osu logo appears!_

While most of the classes aren't working and instead throw errors out(haven't test all of it too), some classes in `osu.Game.Screens` work.

So it lead to my question, Could this be vulnerable? I think it's dangerous to let skin modder use some of the classes that might be dangerous(like editing user data, or overwriting data), or am I just being too aware about this?

**Comment by peppy** [comment]
Thanks for pointing this out!

We'll probably limit it at some point, but I do not believe it can be "dangerous". The only thing you can use it for here is instantiating a class. All further initialisation done on the class is limited, so it couldn't be used for any kind of deserialisation attack.
